Pagination

mariolamacchia opened the pull request.

Commit:
Add pagination on multiple contexts

Commit:
Add context and filter to NUCategory

Commit:
Rename NUCategory method

Commit:
Revert "Add pagination on multiple contexts"

Commit:
Paginate through categories

Commit:
NUModule: add pages to categories

Commit:
Filter categories when paginating

Commit:
NUModule: fix duplicates when paginating

Commit:
Fix pagination when using same fetcher twice

t00f commented on NUModels/NUCategory.j:
```diff
@@ -52,6 +61,14 @@
 
+- (id)initWithName:(CPString)aName contextIdentifier:(CPString)aContext filter:(id)aFilter
+{
+    [self initWithName:aName]
```
This should be the contrary:
`initWithName:` should call `initWithName:contextIdentifier:filter` by setting `contextIdentifier` and `filter` to `nil`

t00f commented on NUModules/NUModule.j:
```diff
@@ -1172,6 +1180,33 @@ NUModuleTabViewModeIcon                = 2;
+- (CPPredicate)_categoryFilterWithFetcher:(NURESTFetcher)aFetcher
+{
+    var userPredicate = [self filter],
+        filter = userPredicate;
```
Weird, no ?

t00f commented on NUModules/NUModule.j:
```diff
@@ -1235,11 +1276,37 @@ NUModuleTabViewModeIcon                = 2;
+    {
+        if (_maxPossiblePage != -1 && _latestPageLoaded >= _maxPossiblePage)
+        {
+            _latestCategoryLoadedIndex++;
```
It looks like we are reseting some values here. Are you sure we need to update the `_latestCategoryLoadedIndex` ?

mariolamacchia replied:
The idea is that when `_latestPageLoaded` reaches `_maxPossiblePage` we have fetched the entire category, so we reset both the values and go to the next one

t00f commented on NUModules/NUModule.j:
```diff
@@ -1235,11 +1276,37 @@ NUModuleTabViewModeIcon                = 2;
+            _latestPageLoaded = -1;
+        }
+
+        if (_latestCategoryLoadedIndex >= [_categories count])
```
I think I would rename `_latestCategoryLoadedIndex` into `_currentPaginatedCategoryIndex`, what do you think ?

mariolamacchia replied:
Yep, looks better

t00f commented on NUModules/NUModule.j:
```diff
@@ -1140,7 +1142,13 @@ NUModuleTabViewModeIcon                = 2;
+    if (_usesPagination && [_categories count])
+    {
+        _latestCategoryLoadedIndex = 0;
+        var context = [_contextRegistry valueForKey:[_categories[0] contextIdentifier]];
```
I would say `[_categories[_latestCategoryLoadedIndex]]`.
I feel `_latestCategoryLoadedIndex` should be renamed

t00f commented on NUModules/NUModule.j:
```diff
@@ -1187,7 +1222,13 @@ NUModuleTabViewModeIcon                = 2;
-    var ID = [aFetcher fetchWithMatchingFilter:[self filter]
+    var filter;
+    if ([_categories count])
+        filter = [self _categoryFilterWithFetcher:aFetcher]
```
Ok good, you are combining the category filter with the current module filter!

t00f replied:
The name of this method is actually confusing.
For now, we have:
```
- (CPPredicate)_categoryFilterWithFetcher:(NURESTFetcher)aFetcher
```
It will not always return a predicate. It can return `nil`, or filter `string`.
I would suggest:
```
- (id)_currentPaginationFilterForFetcher:(NURESTFetcher)aFetcher
```

t00f commented on NUModules/NUModule.j:
```diff
@@ -1140,7 +1142,13 @@ NUModuleTabViewModeIcon                = 2;
         [self showLoading];
 
-    if ([[self class] automaticContextManagement])
+    if (_usesPagination && [_categories count])
```
I am not sure we need `_usesPagination` anymore. I think this can be removed as we will always have pagination. What do you think ?

mariolamacchia replied:
If no context is defined in the category everything should work like before, that means no pagination. In this way we can merge without breaking all the NUModules that use categories.

t00f commented on NUModules/NUModule.j:
```diff
@@ -1172,6 +1180,33 @@ NUModuleTabViewModeIcon                = 2;
+        filter = userPredicate;
+
+    var categoryFilter = [_categories[_latestCategoryLoadedIndex] filter];
+    if (categoryFilter)
```
Prefere the notation that is easier to read:
```
if (!categoryFilter)
    return filter;

// else continue knowing categoryFilter is correctly set
```

t00f commented on NUModules/NUModule.j:
```diff
@@ -3339,7 +3417,16 @@ NUModuleTabViewModeIcon                = 2;
+            _usesPagination = NO;
+            break;
+        }
+    }
```
I think we should force the category to have a fetcher, no ?
If we agree, let's throw an error to ensure all categories have a fetcher and we will fix our current code

mariolamacchia replied:
If we go for it we'll need to change all the controllers with categories, isn't that a big change?

t00f replied:
I found 56 results, which is not a big deal. It will be part of the PR we will suggest on our repository

t00f reviewed (changes requested):
@mariolamacchia This looks like a good start. Let's ensure everythin is working fine as this library is part of our core solution.

I have added few comments. We can discuss them offline if you want

Commit:
Fix NUCategory constructors

Commit:
Replace number with variable

Commit:
Rename attribute name

Commit:
NUModule: refactor filter building

t00f commented on NUModels/NUCategory.j:
```diff
@@ -51,24 +51,24 @@
 
 - (id)initWithName:(CPString)aName
 {
+    [self initWithName:aName contextIdentifier:nil filter:nil];
```
```return [self initWithName:aName contextIdentifier:nil filter:nil];```
should be enough.

Also, if we go for a mandatory context, we will need to update this constructor

Commit:
Implement new pagination in NUAbstractObjectAssociator

chris-serafin committed:
Small change to make things clear

chris-serafin committed:
Merge branch 'pagination' of https://github.com/mariolamacchia/nukit into pagination

Commit:
Load next page if last was not full

Commit:
Fix duplicates when clearing filter

Commit:
Revert "Fix duplicates when clearing filter"

This reverts commit 1118841376530b50177510bf128fb1b6acc2d9da.

Commit:
Fix duplicates results and requests

This fixes:

- duplicate elements in wrong category after clearing the search box
- duplicate requests when typing in the search box

The issue is that the method `_flushCategoriesContent` is causing
`observeValueForKeyPath:ofObject:change:context` to be fired. If we remove
the scrollview observer everything works fine (it is reapplied after fetching)

Commit:
Add workaround to missing API filters

The easiest way was to call `categoryForObject` and give it priority over the
current category. The object will be assigned to the currentCategory only if
the return value is nil: in this way both new and old modules will keep working

Commit:
Fix refetching context

After the previous commit, if two categories shared context and filter, the
requests were made even though no change was visible to the user

Commit:
Add checks for dynamically added categories

chris-serafin committed:
Improved syntax
